### PR TITLE
Implement LTA lifecycle commands

### PR DESCRIPTION
**Pull request by blinkdog**
By the magic of semaphores, we can now start and stop components with `ltacmd`.

Commands are:
`ltacmd drain` - Drain the currently running component, then stop
`ltacmd start <component>` - Start the named component
`ltacmd stop` - Stop the named component

Bugs are:
`ltacmd kill` - We save the PID generated by `ltacmd start`, but `kill -9` only gets the bash script, not the Python process. Still researching a way to make this work nicely.

### CODE REVIEW

File: lta/bundler.py [hunk 1]
```diff
@@ -233,6 +235,21 @@ def __init__(self, config: Dict[str, str], logger: Logger) -> None:
         }
         await lta_rc.request('POST', '/Files/actions/bulk_update', update_body)
 
+
+def check_drain_semaphore() -> bool:
+    """Check if a drain semaphore exists in the current working directory."""
+    cwd = os.getcwd()
+    drain_filename = os.path.join(cwd, DRAIN_SEMAPHORE_FILENAME)
+    return Path(drain_filename).exists()
```
Comment (dsschult) [inline lta/bundler.py]: This function seems like duplicate code, since it's in every module.

File: lta/bundler.py [hunk 2]
```diff
@@ -233,6 +235,21 @@ def __init__(self, config: Dict[str, str], logger: Logger) -> None:
...
+
+
+def check_stop_semaphore() -> bool:
+    """Check if a stop semaphore exists in the current working directory."""
+    cwd = os.getcwd()
+    drain_filename = os.path.join(cwd, STOP_SEMAPHORE_FILENAME)
+    return Path(drain_filename).exists()
```
Comment (dsschult) [inline lta/bundler.py]: As I point out later, this function seems unnecessary.

File: lta/bundler.py [hunk 3]
```diff
@@ -265,10 +282,25 @@ def __init__(self, config: Dict[str, str], logger: Logger) -> None:
...
+async def lifecycle_loop(bundler: Bundler) -> None:
+    """Run a check for a stop semaphore as an infinite loop."""
+    bundler.logger.info("Starting lifecycle loop")
+    while True:
+        # if there is a stop semaphore, terminate the program
+        if check_stop_semaphore():
+            bundler.logger.info("Component stopped; shutting down.")
+            sys.exit(0)
+        # sleep until we check again
+        await asyncio.sleep(1.0)
```
Comment (dsschult) [inline lta/bundler.py]: Might be easier to just get a SIGTERM instead of checking explicitly for something every second.

File: lta/bundler.py [hunk 4]
```diff
@@ -265,10 +282,25 @@ def __init__(self, config: Dict[str, str], logger: Logger) -> None:
...
+
+
 async def status_loop(bundler: Bundler) -> None:
     """Run status heartbeat updates as an infinite loop."""
     bundler.logger.info("Starting status loop")
     while True:
+        # if there is a drain semaphore, stop sending status updates
+        if check_drain_semaphore():
+            break
```
Comment (dsschult) [inline lta/bundler.py]: Could you just do:

```python
while not check_drain_semaphore():
```

File: lta/bundler.py [hunk 5]
```diff
@@ -279,6 +311,10 @@ def __init__(self, config: Dict[str, str], logger: Logger) -> None:
     """Run bundler work cycles as an infinite loop."""
     bundler.logger.info("Starting work loop")
     while True:
+        # if there is a drain semaphore, don't do any additional work
+        if check_drain_semaphore():
+            bundler.logger.info("Component drained; shutting down.")
+            sys.exit(0)
```
Comment (dsschult) [inline lta/bundler.py]: Same as above, putting it in the `while` condition.

File: lta/lta_cmd.py [hunk 6]
```diff
@@ -191,17 +288,32 @@ def date_ok(d: str) -> bool:
...
                                help="display output in JSON",
                                action="store_true")
     parser_status.set_defaults(func=status)
 
+    # define a subparser for the 'stop' subcommand
+    parser_stop = subparser.add_parser('stop', help='stop the component')
+    parser_stop.add_argument("component",
+                             choices=COMPONENT_NAMES,
+                             help="optional LTA component",
+                             nargs='?')
```
Comment (dsschult) [inline lta/lta_cmd.py]: If we're just going to ignore the component name, why take it as an argument?

File: lta/bundler.py (hunk 1 above)
Comment (blinkdog) [reply lta/bundler.py]: This is one that will probably end up refactored in DEV-605:
https://wipacit.atlassian.net/browse/DEV-605

File: lta/bundler.py (hunk 2 above)
Comment (blinkdog) [reply lta/bundler.py]: Yep, this one went away.

File: lta/bundler.py (hunk 3 above)
Comment (blinkdog) [reply lta/bundler.py]: Yep, this too also went away.

File: lta/bundler.py (hunk 4 above)
Comment (blinkdog) [reply lta/bundler.py]: I liked this better, so I did end up using it verbatim.

File: lta/bundler.py (hunk 5 above)
Comment (blinkdog) [reply lta/bundler.py]: Yep.

File: lta/lta_cmd.py [hunk 7]
```diff
@@ -31,6 +33,33 @@
 }
 
 
+def deamon_for(component: str) -> Daemon:
+    """Create a Daemon object for the specified component name."""
+    pidfile = pid_filename(component)
+    runner = __import__(f"lta.{component}", fromlist=['']).runner
```
Comment (dsschult) [inline lta/lta_cmd.py]: Should use `importlib.import_module`.

File: lta/lta_cmd.py (hunk 6 above)
Comment (blinkdog) [reply lta/lta_cmd.py]: I ended up making this component argument non-optional.
On the upside, multiple components can be started from the same repo directory.

File: lta/lta_cmd.py (hunk 7 above)
Comment (blinkdog) [reply lta/lta_cmd.py]: Fixed this one up.

**Pull request merged**